rpcs3/Emu/RSX/rsx_methods.cpp

```diff
@@ -450,17 +451,37 @@ namespace rsx
 		{
 			static void impl(thread* rsx, u32 _reg, u32 arg)
 			{
-				if (rsx::method_registers.transform_program_load() >= 512)
+				// Get real args count
+				const u32 count = std::min<u32>({rsx->fifo_ctrl->get_remaining_args_count() + 1,
+					static_cast<u32>(((rsx->ctrl->put & ~3ull) - (rsx->fifo_ctrl->get_pos() - 4)) / 4), 32 - index});
+
+				const u32 load_pos = rsx::method_registers.transform_program_load();
+
+				u32 rcount = count;
+
+				if (const u32 max = load_pos * 4 + rcount + (index % 4);
+					max > 512 * 4)
 				{
 					// PS3 seems to allow exceeding the program buffer by upto 32 instructions before crashing
 					// Discard the "excess" instructions to not overflow our transform program buffer
 					// TODO: Check if the instructions in the overflow area are executed by PS3
 					rsx_log.warning("Program buffer overflow!");
-					return;
+					rcount -= max - (512 * 4);
 				}
 
-				method_registers.commit_4_transform_program_instructions(index);
-				rsx->m_graphics_state |= rsx::pipeline_state::vertex_program_dirty;
+				alignas(16) u8 buffer[128];
+				stream_data_to_memory_swapped_u32(buffer, vm::base(rsx->fifo_ctrl->get_current_arg_ptr()), rcount, 4);
+
+				const auto dst = &rsx::method_registers.transform_program[load_pos * 4 + index % 4];
+
+				if (std::memcmp(dst, buffer, rcount * 4) != 0)
+				{
+					rsx->m_graphics_state |= rsx::pipeline_state::vertex_program_dirty;
```

Comment:
I have mixed feelings on this one. The vertex program block does not actually contain one vertex program, its a memory pool like any other. i.e This update may have nothing to do with the current vertex program and will just lead to false positives.
Unfortunately, determining the range of a vertex shader is not easy with runtime patching being used to stitch together separate subroutines. I'd suggest to just stub this with a call to a function `RSXThread::invalidate_vertex_program(range)` and we can go from there. It's a lot more complicated than just checking for changed bytes.